dcpy/connectors/socrata/publish.py

```diff
@@ -0,0 +1,350 @@
+                        "attachments": attachments,
+                        "metadata": metadata,
+                    }
+                ),
+            )
+        )
+
+    def push_shp(
+        self, path: Path, *, dest_filename: str | None = None
+    ) -> Socrata.Responses.RevisionDataSource:
+        # This is the only use of socrata-py. The data syncing utilities are nice.
+        _socratapy_client = SocrataPy(
+            Authorization(SOCRATA_DOMAIN, SOCRATA_USER, SOCRATA_PASSWORD)
+        )
+        view = _socratapy_client.views.lookup(self.four_four)
+        rev = view.revisions.lookup(self.revision_num)
+        with open(path, "rb") as shp_zip:
+            logger.info(
+                f"Pushing shapefiles at {path} to {self.four_four} - rev: {self.revision_num}"
+            )
+            push_resp = (
+                rev.create_upload(dest_filename or shp_zip.name)
+                .shapefile(shp_zip)
+                .wait_for_finish()
+            )
+        error_details: dict | None = push_resp.attributes["failure_details"]
+        if error_details:
+            logger.error(f"Shapefile upload failed with {error_details}")
+            raise Exception()
+        return Socrata.Responses.RevisionDataSource(push_resp)
+
+    def upload_attachment(
+        self, path: Path, *, file_name: str | None = None
+    ) -> Socrata.Inputs.Attachment:
+        """Upload an attachment to a revision.
+
+        Note: uploading an attachment isn't enough to truly attach it to a revision.
+        It must also be patched into the metadata revision. Unfortunately, subsequent
+        patches seem to wipe out the attachment metadata, everything must be gathered
+        up into one patch.
+        """
+        # TODO: check other attachment types, e.g. xlsx
+        content_types = {".csv": "text/csv", ".pdf": "application/pdf"}
+
+        with open(path, "rb") as f:
+            attachment_md_raw = _socrata_request(
+                f"{self.revision_endpoint}/attachment",
+                "POST",
+                headers={
+                    "X-File-Name": file_name or path.name,
+                    "Content-type": content_types[path.suffix],
+                },
+                data=f,
+            )
+        return Socrata.Inputs.Attachment(
+            # TODO: diff between name and filename? and what about blob_id?
+            name=attachment_md_raw["filename"],
+            filename=attachment_md_raw["filename"],
+            asset_id=attachment_md_raw["file_id"],
+            blob_id=None,
+        )
+
+
+def push_shp_from_s3(dataset_name: str, version: str):
+    """Push Shapefile from edm-packaging to Socrata, and sync metadata."""
+    dataset_metadata: dict = datasets[dataset_name]
+    four_four: str = dataset_metadata["four_four"]
+    socrata_metadata = dataset_metadata["metadata"]
+    logger.info(
+        f"Syncing shapefile and metadata for {dataset_name}/{version} to Socrata {four_four}"
+    )
```

Comment:
might be nice to have the full `four_four` URL to navigate to either at the start or end of a sync, maybe just the end to see the changes on Open Data

Reply:
that's a great idea. We can definitely add a link to the dataset in socrata. And maybe if the revision is left hanging, a link to that as well.